### PR TITLE
plan, executor: improve index join to support all join types

### LINKED ISSUE #4901: improve index join to support all join types

### PR DESCRIPTION
**Pull request by zz-jason**
fix #4901

### CODE REVIEW

**Comment by zz-jason** [comment]
@hanfei1991 @winoros PTAL

File: plan/physical_plans.go [hunk 1]
```diff
@@ -1173,8 +1173,18 @@ func buildSchema(p PhysicalPlan) {
 	switch x := p.(type) {
 	case *Limit, *TopN, *Sort, *Selection, *MaxOneRow, *SelectLock:
 		p.SetSchema(p.Children()[0].Schema())
-	case *PhysicalHashJoin, *PhysicalIndexJoin:
+	case *PhysicalHashJoin:
 		p.SetSchema(expression.MergeSchema(p.Children()[0].Schema(), p.Children()[1].Schema()))
+	case *PhysicalIndexJoin:
```
Comment (hanfei1991) [inline plan/physical_plans.go]: This logic is the same as MergeJoin ?

File: plan/new_physical_plan_builder.go [hunk 2]
```diff
@@ -108,26 +108,38 @@ func joinKeysMatchIndex(keys []*expression.Column, index *model.IndexInfo) []int
...
 		DefaultValues:   p.DefaultValues,
 		outerSchema:     p.children[outerIdx].Schema(),
 		innerPlan:       innerPlan,
 	}.init(p.allocator, p.ctx, p.children[outerIdx], p.children[1-outerIdx])
-	join.SetSchema(expression.MergeSchema(p.children[outerIdx].Schema(), p.children[1-outerIdx].Schema()))
+	switch p.JoinType {
+	case SemiJoin, AntiSemiJoin:
+		join.SetSchema(p.children[0].Schema().Clone())
+	case LeftOuterSemiJoin, AntiLeftOuterSemiJoin:
+		join.SetSchema(p.children[0].Schema().Clone())
```
Comment (hanfei1991) [inline plan/new_physical_plan_builder.go]: `join.SetSchema(p.Schema().Clone())` ?

File: executor/index_reader.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
+// Copyright 2017 PingCAP, Inc.
```
Comment (hanfei1991) [inline executor/index_reader.go]: Is it ok to put `index_reader` and `index_lookup` together ? It's confusing with `index_lookup` and `index_lookup_join`.

File: plan/physical_plans.go (hunk 1 above)
Comment (zz-jason) [reply plan/physical_plans.go]: No, there is a little difference when the join type is `InnerJoin`, `LeftOuterJoin` and `RightOuterJoin`

File: executor/index_reader.go (hunk 3 above)
Comment (zz-jason) [reply executor/index_reader.go]: reverted

File: plan/new_physical_plan_builder.go (hunk 2 above)
Comment (zz-jason) [reply plan/new_physical_plan_builder.go]: done

**Comment by hanfei1991** [comment]
LGTM

**Comment by zz-jason** [comment]
@coocood PTAL

File: executor/index_lookup_join.go [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+
+	resultGenerator joinResultGenerator
+	resultBuffer    []Row
+	resultCursor    int
+
+	buffer4JoinKeys [][]types.Datum
+	buffer4JoinKey  []types.Datum
+
+	batchSize int
+	exhausted bool // exhausted means whether all data has been extracted
```
Comment (winoros) [inline executor/index_lookup_join.go]: add `.`

File: executor/index_lookup_join.go [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+type IndexLookUpJoin struct {
+	baseExecutor
+
+	outerExec        Executor
+	innerExec        DataReader
+	outerKeys        []*expression.Column
+	innerKeys        []*expression.Column
+	outerFilter      expression.CNFExprs
+	innerFilter      expression.CNFExprs
+	outerOrderedRows *keyRowBlock
```
Comment (winoros) [inline executor/index_lookup_join.go]: `outerRowBlock` is better? It's not ordered when initialized.

File: executor/index_lookup_join.go [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+	e.resultGenerator = nil
+	e.resultBuffer = nil
+	e.buffer4JoinKeys = nil
+	e.buffer4JoinKey = nil
+
+	return nil
+}
+
+// Next implements the Executor Next interface.
+// step1: fetch a batch of "outer rows".
```
Comment (winoros) [inline executor/index_lookup_join.go]: step -> Step

File: executor/index_lookup_join.go [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+		key, err := codec.EncodeKey(nil, joinKey...)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		keys = append(keys, key)
+	}
+	return keys, nil
+}
+
+// requestKeys must be ordered.
```
Comment (winoros) [inline executor/index_lookup_join.go]: Make the comment more formal.

File: executor/index_lookup_join.go (hunk 5 above)
Comment (zz-jason) [reply executor/index_lookup_join.go]: I think `outerOrderedRows` is better, it means we should make it ordered before using it.

**Pull request merged**